### ISSUE TITLE
Zelig and Tidyverse

### ISSUE THREAD

**Issue by christophergandrud** [IQSS/Zelig#222]
@mercecrosas following on from our meeting this morning, in 578171794494d9e5ea2954d05570886058398f28 I put a minimal vignette in flagging that Zelig accepts tidy data and can be 'piped', e.g.:

```r
swiss %>% 
    zelig(Fertility ~ Agriculture + Examination, model = 'ls', data = ., 
          cite = FALSE) %>%
    setx(Agriculture = seq(1, 90, by = 5)) %>%
    sim() %>% 
    plot()
``` 

The Zelig workflow makes a lot of sense in piped workflow. 

Pipes don't work with the reference class syntax, e.g.

```r
zls$new() %>%
zls$new() %>%
  .$zelig(Fertility ~ Education, data = swiss)

 ## Error in as.vector(x, "character") : 
 ##  cannot coerce type 'environment' to vector of type 'character' 
```

Another issue is creating first differences as calling the `setx` wrapper twice just overwrites results from the first call. There is no wrapper for `setx1`. So, to improving piping it might be worthwhile to create a wrapper for `setx1`.

Finally, introducing the `zelig_qi_to_df` will allow users to make ggplot2 plots, e.g. https://github.com/IQSS/Zelig/issues/189.

**Comment by cchoirat** [comment]
Definitely worth exploring.

To make pipes & reference classes work together, you may want to try
solutions along the lines of:

library(Zelig)

library(magrittr)

data(cars)

Zelig <- zelig

Setx <- function(obj, ...) { obj$setx(...); return(obj) }

Setx1 <- function(obj, ...) { obj$setx1(...); return(obj) }

Sim <- function(obj, ...) { obj$sim(...); return(obj) }

Plot <- function(obj, ...) { obj$graph(...); return(obj)}

Summary <- function(obj) { summary(obj); return(obj) }

set.seed(1)

cars %>%

  Zelig(dist ~ speed, data = ., model = "ls")  %>%

  Summary %>%

  Setx(speed = 18) %>%

  Sim %>%

  Summary -> z.out



which gives


Call:
> z5$zelig(formula = formula, data = data, by = by)
> Residuals:
>     Min      1Q  Median      3Q     Max
> -29.069  -9.525  -2.272   9.215  43.201
> Coefficients:
>             Estimate Std. Error t value Pr(>|t|)
> (Intercept) -17.5791     6.7584  -2.601   0.0123
> speed         3.9324     0.4155   9.464 1.49e-12
> Residual standard error: 15.38 on 48 degrees of freedom
> Multiple R-squared:  0.6511, Adjusted R-squared:  0.6438
> F-statistic: 89.57 on 1 and 48 DF,  p-value: 1.49e-12
> Next step: Use 'setx' method
>  sim x :
>  -----
> ev
>       mean       sd      50%    2.5%    97.5%
> 1 53.23961 2.523266 53.32596 48.2635 58.17909
> pv
>          mean       sd      50%     2.5%    97.5%
> [1,] 53.47505 16.01821 52.79777 20.85769 83.58387
>

**Comment by christophergandrud** [comment]
Looks good, I think we basically have all of this apart from a `setx1` function using the Zelig 4 wrappers.

**Comment by christophergandrud** [comment]
Added the `setx1` wrapper in c504676a4bef596860ff94310684676d9e6c1ead

So you can now do:

```r
library(Zelig)
library(dplyr)
data(turnout)

# plot first differences
zelig(Fertility ~ Education, data = swiss, model = 'ls') %>%
      setx(z4, Education = 10) %>%
      setx1(z4, Education = 30) %>%
      sim() %>%
      plot()
```

**Comment by christophergandrud** [comment]
`zelig_qi_to_df` implemented in e666b475c0a493aa4acef4b1511c325dde06030e

For example:

```r
zelig(Petal.Width ~ Petal.Length + Species, 
                             data = iris, model = "ls") %>%
  setx(Petal.Length = 2:4, Species = "setosa") %>%
  sim() %>%
  zelig_qi_to_df() %>% 
  qi_slimmer()
```

returns:

```r
## x Petal.Length Speciesversicolor Speciesvirginica    qi_min qi_median    qi_max
## 1 x            2                 0                0                            0.3112097 0.3706236 0.4295065
## 2 x            3                 0                0                            0.4851957 0.6028071 0.7159249
## 3 x            4                 0                0                           0.6520305 0.8328370 1.0173164
```

`zelig_qi_to_df` returned the entire set of simulations and `qi_slimmer cuts it down to just the central interval for each simulation (95% by default).

**Comment by christophergandrud** [comment]
. . . example added in 9e8d21c1511d2a7ecb6506cbbb3b56ff88856c3c @christophergandrud christophergandrud committed 25 seconds ago
